### PR TITLE
Blog post about a RAG application using Granite

### PR DESCRIPTION
**Pull request by cescoffier**

### CODE REVIEW

File: _posts/2024-06-14-granite-rag.adoc [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+=== Retrieval
+
+The second step is the retrieval process.
+When a user inputs a query, the application searches the vector database to find the most relevant text segments.
+
+To achieve this, the application computes the vector representation of the user query using the **same** embedding model and compares it with the vector representations of the stored text segments in Infinispan.
+It selects the most relevant text segments based on the similarity between the query vector and the text segment vectors.
+
+Then, it augments the user query with the retrieved text segments and sends it to the AI Service.
```
Comment (geoand) [inline _posts/2024-06-14-granite-rag.adoc:74]:
```suggestion
Then, it augments the user query with the retrieved text segments and sends it to the LLM
```

File: _posts/2024-06-14-granite-rag.adoc [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+* `quarkus-web-bundler`: Bundles frontend resources with the Quarkus application.
+
+We also need a specific dependency to use the BGE-small embedding model:
+
+[source,xml]
+----
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j-embeddings-bge-small-en-q</artifactId>
+    <version>0.31.0</version>
```
Comment (geoand) [inline _posts/2024-06-14-granite-rag.adoc:98]: You don't the version if you use our BOM

File: _posts/2024-06-14-granite-rag.adoc [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+# Configure the OpenAI service to use instruct lab:
+quarkus.langchain4j.openai.base-url=http://localhost:8000/v1
+quarkus.langchain4j.openai.timeout=60s
+
+quarkus.langchain4j.embedding-model.provider=dev.langchain4j.model.embedding.BgeSmallEnQuantizedEmbeddingModel
+----
+
+The first line configures the dimension of the vectors stored in Infinispan, which depends on the embedding model (BGE-small in our case).
+The second and third lines configure the OpenAI service to use Instructlab.
+You can replace the base URL with the one for Podman AI Studio if you prefer.
```
Comment (geoand) [inline _posts/2024-06-14-granite-rag.adoc:120]: I think we should explain here that these services implement the JSON API that OpenAI SaaS defines and that is why we can do this

File: _posts/2024-06-14-granite-rag.adoc [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+synopsis: 'How to build a RAG chatbot using the Granite LLM.'
+author: cescoffier
+---
+:imagesdir: /assets/images/posts/llms
+
+This blog post demonstrate how to build an AI-infused chatbot application using Quarkus, Langchain4J, Infinispan, and the Granite LLM.
+In this post, we will create an **entirely** local solution, eliminating the need for any cloud services, including the LLM.
+
+Our chatbot leverages the Granite LLM, a language model that generates contextually relevant text based on user prompts.
+To run Granite locally, we'll be utilizing Instructlab, though Podman AI Studio is another viable option.
```
Comment (gsmet) [inline _posts/2024-06-14-granite-rag.adoc:15]: I fixed only one, there might be other occurrences.
```suggestion
To run Granite locally, we'll be utilizing InstructLab, though Podman AI Studio is another viable option.
```

File: _posts/2024-06-14-granite-rag.adoc [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+== Architecture
+
+Our chatbot application is composed of four main components:
+
+1. **Web Socket Endpoint**: This component serves as the communication bridge between the chatbot's backend and the frontend interface.
+This component uses the new Quarkus WebSocket-Next extension to handle WebSocket connections efficiently.
+It relies on the AI Service to interact with the LLM.
+
+2. **Ingestor**: The ingestor is responsible for populating the database with relevant data. It processes a set of local document, split them into text segment, compute their vector representation and store them into Infinispan.
```
Comment (gsmet) [inline _posts/2024-06-14-granite-rag.adoc:29]:
```suggestion
2. **Ingestor**: The ingestor is responsible for populating the database with relevant data. It processes a set of local documents, split them into text segments, compute their vector representation and store them into Infinispan.
```

File: _posts/2024-06-14-granite-rag.adoc [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+3. **Retriever**: The retriever allows finding relevant text segment into Infinispan. When a user inputs a query, the retriever searches the vector database to find the most relevant pieces of information.
```
Comment (gsmet) [inline _posts/2024-06-14-granite-rag.adoc:31]:
```suggestion
3. **Retriever**: The retriever allows finding relevant text segments into Infinispan. When a user inputs a query, the retriever searches the vector database to find the most relevant pieces of information.
```

File: _posts/2024-06-14-granite-rag.adoc [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+
+There are more advanced version of the RAG pattern, but let's stick to the basics for this application.
+
+=== Ingestion
+
+Let's first look at the ingestion step.
+The ingestion process involves reading a set of documents, splitting them into text segments, computing their vector representations, and storing them in Infinispan.
+
+The secret of an effective RAG implementation lies in how the text segments are computed.
+In our application, we will follow a straightforward approach, but more advanced techniques are available and often required.
```
Comment (gsmet) [inline _posts/2024-06-14-granite-rag.adoc:60]: Do you have pointers to these more advanced techniques? Might be interesting to add links.

File: _posts/2024-06-14-granite-rag.adoc [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+quarkus.langchain4j.openai.timeout=60s
+
+quarkus.langchain4j.embedding-model.provider=dev.langchain4j.model.embedding.BgeSmallEnQuantizedEmbeddingModel
+----
+
+The first line configures the dimension of the vectors stored in Infinispan, which depends on the embedding model (BGE-small in our case).
+The second and third lines configure the OpenAI service to use Instructlab.
+You can replace the base URL with the one for Podman AI Studio if you prefer.
+
+The last line sets the default embedding model to BGE-small.
```
Comment (gsmet) [inline _posts/2024-06-14-granite-rag.adoc:122]: Maybe use callouts instead?

File: _posts/2024-06-14-granite-rag.adoc [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+
+=== The Ingestor
+
+With the configuration in place, let's implement the https://github.com/cescoffier/quarkus-granite-rag-demo/blob/main/src/main/java/me/escoffier/granite/rag/Ingestion.java[ingestor].
+The ingestor reads documents from the documents directory, splits them into text segments, computes their vector representations, and stores them in Infinispan:
```
Comment (gsmet) [inline _posts/2024-06-14-granite-rag.adoc:127]:
```suggestion
The ingestor reads documents from the `documents` directory, splits them into text segments, computes their vector representations, and stores them in Infinispan:
```

File: _posts/2024-06-14-granite-rag.adoc [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+    public Ingestion(EmbeddingStore<TextSegment> store, EmbeddingModel embedding) {
+
+        EmbeddingStoreIngestor ingestor = EmbeddingStoreIngestor.builder()
+                .embeddingStore(store)
+                .embeddingModel(embedding)
+                .documentSplitter(recursive(1024, 0))
+                .build();
+
+        File file = new File("documents");
+        List<Document> documents = FileSystemDocumentLoader.loadDocuments(file.toPath());
```
Comment (gsmet) [inline _posts/2024-06-14-granite-rag.adoc:144]: Maybe use a `Path` directly?

File: _posts/2024-06-14-granite-rag.adoc [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+More advanced techniques may be used to improve the accuracy of the RAG model.
+
+=== The retriever
+
+Next, let's implement the https://github.com/cescoffier/quarkus-granite-rag-demo/blob/main/src/main/java/me/escoffier/granite/rag/Retriever.java[retriever].
+The retriever finds the most relevant text segments in Infinispan based on the user query:
+
+[source,java]
+----
+@Singleton
```
Comment (gsmet) [inline _posts/2024-06-14-granite-rag.adoc:168]: From our experience, it's better to provide the imports.

File: _posts/2024-06-14-granite-rag.adoc [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+}
+----
+
+The `@RegisterAiService` annotation specifies the retrieval augmentor to use, which in our case is the `Retriever` bean defined earlier.
+
+The `@SystemMessage` annotation provides the main instructions for the AI model.
+
+The `@SessionScoped` annotation ensures that the AI service is stateful, maintaining context between user interactions for more engaging conversations.
+
+The `ChatBot` interface defines a single method, `chat`, which takes a user question as input and returns the chatbot's response.
```
Comment (gsmet) [inline _posts/2024-06-14-granite-rag.adoc:225]: Same here, callouts would be better?

File: _posts/2024-06-14-granite-rag.adoc [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+    String onMessage(String message) {
+        return bot.chat(message);
+    }
+
+}
+----
+
+The `@WebSocket` annotation specifies the WebSocket path.
+The `@OnOpen` method sends a welcome message when a user connects to the _WebSocket_.
+The `@OnTextMessage` method processes the user's messages and returns the chatbot's responses, using the injected AI service.
```
Comment (gsmet) [inline _posts/2024-06-14-granite-rag.adoc:255]: Callouts?

File: _posts/2024-06-14-granite-rag.adoc [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+That's it! Our chatbot is now ready to chat with users, providing contextually relevant responses based on the RAG pattern.
+
+== Running the application
+
+Let's run the application and see our chatbot in action.
+First, clone the https://github.com/cescoffier/quarkus-granite-rag-demo/tree/main[repository] and run the following command:
+
+[source,shell]
+----
+mvn quarkus:dev
```
Comment (gsmet) [inline _posts/2024-06-14-granite-rag.adoc:266]:
```suggestion
./mvnw quarkus:dev
```

File: _posts/2024-06-14-granite-rag.adoc [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+
+image::mona-screenshot.png[The Mon chatbot,float="right",align="center"]
+
+== Summary
+
+That's it!
+With just a few lines of code, we have implemented a chatbot using the RAG pattern, combining the capabilities of the Granite LLM, Infinispan, and Quarkus.
+This application runs entirely locally, eliminating the need for any cloud services and addressing privacy concerns.
+
+This is just an example of what you can achieve with the Quarkus Langchain4j extension.
```
Comment (gsmet) [inline _posts/2024-06-14-granite-rag.adoc:284]:
```suggestion
This is just an example of what you can achieve with the Quarkus LangChain4j extension.
```

File: _posts/2024-06-14-granite-rag.adoc [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+
+Once the application is up and running, open your browser and navigate to http://localhost:8080.
+You should see the chatbot interface, where you can start chatting with Mona:
+
+image::mona-screenshot.png[The Mon chatbot,float="right",align="center"]
+
+== Summary
+
+That's it!
+With just a few lines of code, we have implemented a chatbot using the RAG pattern, combining the capabilities of the Granite LLM, Infinispan, and Quarkus.
```
Comment (gsmet) [inline _posts/2024-06-14-granite-rag.adoc:281]: I think at some point, I would make it clear than a lot more models/document stores are available in the Quarkus LangChain4j project if you prefer to use something else.

**Review by gsmet — COMMENTED** [review]
I added some comments inline.

File: _posts/2024-06-14-granite-rag.adoc (hunk 2 above)
Comment (cescoffier) [reply _posts/2024-06-14-granite-rag.adoc:98]: So, code.quarkus.io does not add the bom (that's why I have the version) - let me remove it from the blog post but keep it in the code.

File: _posts/2024-06-14-granite-rag.adoc (hunk 7 above)
Comment (cescoffier) [reply _posts/2024-06-14-granite-rag.adoc:60]: Added in the text.

File: _posts/2024-06-14-granite-rag.adoc (hunk 11 above)
Comment (cescoffier) [reply _posts/2024-06-14-granite-rag.adoc:167]: I've linked the full class.

**Review by geoand — APPROVED** [review]

**Review by gsmet — APPROVED** [review]

File: _posts/2024-06-14-granite-rag.adoc [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+Let's run the application and see our chatbot in action.
+First, clone the https://github.com/cescoffier/quarkus-granite-rag-demo/tree/main[repository] and run the following command:
+
+[source,shell]
+----
+./mvnw quarkus:dev
+----
+
+This command starts the Quarkus application in development mode.
+Ensure you have InstructLab or Podman AI Studio running to use the Granite LLM.
```
Comment (maxandersen) [inline _posts/2024-06-14-granite-rag.adoc:260]: should we mention on windows and mac better to use instructlab due to lack of GPU utilization in containers ?

**Pull request merged**